### ISSUE TITLE
Instrumentation hook is not called in `standalone` output

### ISSUE THREAD

**Issue by madebyherzblut** [vercel/next.js#49897]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.4.0: Mon Mar  6 21:00:17 PST 2023; root:xnu-8796.101.5~3/RELEASE_X86_64
    Binaries:
      Node: 18.16.0
      npm: 9.5.1
      Yarn: 1.22.19
      pnpm: 8.5.0
    Relevant packages:
      next: 13.4.3-canary.1
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

Standalone mode (output: "standalone")

### Link to the code that reproduces this issue

https://github.com/madebyherzblut/next-instrumentation-standalone

### To Reproduce

1. Create a new Next project using `create-next-app`
2. Enable the `instrumentationHook` and `standalone` output in `next.config.js`
3. Build the project using `next build`
4. Run the server with `.next/standalone/server.js`

### Describe the Bug

The instrumentation hook is executed with `next dev`, but not in `standalone` mode. Based on the example provided in https://github.com/vercel/next.js/issues/48262 I could trace the issue back to changes between `next@13.3.5-canary.8` and `next@13.3.5-canary.9`. If the hook is called it should output "Instrumenting nodejs runtime" in the examples below:

```sh
$ pnpm add next@13.3.5-canary.8
[…]
$ NODE_ENV=production pnpm run build && node .next/standalone/server.js

[…]

Listening on port 3000 url: http://localhost:3000
Instrumenting nodejs runtime                                                # Correct
                                                                                                                                                                                     
$ rm -Rf .next node_modules
$ pnpm add next@13.3.5-canary.9
[…]
$ NODE_ENV=production pnpm run build && node .next/standalone/server.js

[…]

Listening on port 3000 url: http://localhost:3000                 # Missing output from hook
                                                                                                                                                                                      
$ pnpm run dev

[…]

event - compiled client and server successfully in 256 ms (326 modules)
Instrumenting nodejs runtime

$ pnpm add next@canary
Packages: +4 -3
++++---

dependencies:
- next 13.3.5-canary.9
+ next 13.4.3-canary.1

Progress: resolved 28, reused 20, downloaded 0, added 4, done
Done in 3.7s
$ NODE_ENV=production pnpm run build && node .next/standalone/server.js

[…]

Listening on port 3000 url: http://localhost:3000             # Missing output from hook
                                                                                         
$ pnpm run dev

> next-instrumentation-standalone@0.1.0 dev /Users/chris/Development/next-instrumentation-standalone
> next dev

- ready started server on 0.0.0.0:3000, url: http://localhost:3000
- warn You have enabled experimental feature (instrumentationHook) in next.config.js.
- warn Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use at your own risk.

- event compiled client and server successfully in 1770 ms (304 modules)
- wait compiling...
- wait compiling /instrumentation (client and server)...
- event compiled client and server successfully in 287 ms (319 modules)
Instrumenting nodejs runtime                                            # Correct in `dev` mode
```

### Expected Behavior

The instrumentation hook should be executed in `standalone` outputs.

**Comment by tomsseisums** [comment]
Works for me on `v13.4.2`. The hook seems to not be called on launching the server, but on handling each request:
```
$ PORT=4200 node .next/standalone/server.js
- info Loaded env from [...]/.next/standalone/.env
Listening on port 4200 url: http://localhost:4200
register instrumentation: nodejs
- info Loaded env from [...]/.next/standalone/.env
register instrumentation: nodejs
register instrumentation: edge
- info Loaded env from [...]/.next/standalone/.env
register instrumentation: nodejs
```

**Comment by madebyherzblut** [comment]
Thanks @tomsseisums, I can confirm that it runs when handling a request. When I read the documentation, I thought there is only one server instance. So is this expected behavior?

> If you export a function named register from this file, we will call that function whenever a new Next.js server instance is bootstrapped. When your register function is deployed, it will be called on each cold boot (but exactly once in each environment).
https://nextjs.org/docs/app/building-your-application/optimizing/instrumentation

**Comment by 1148177968** [comment]
I had some problems, too.
register is called twice in dev mode and not at all in production mode (Instance startup time).
And when I only import or require server-side modules from within register, it still complains about server-side modules not being available (e.g. path and fs). I've already import it dynamically.

**Comment by 1148177968** [comment]
i have the same problem

**Comment by maliesa96** [comment]
same problem as well

**Comment by jviall** [comment]
I believe I'm also experiencing this. I use `standalone` mode; `register()` function is called when using `next dev`, but when I deploy to Vercel, there's no sign of it ever running. Using Next `13.4.19`

**Comment by MeghdadHadidi** [comment]
Has anyone found any workaround for this? I've already tried canary version, no chance. 
I've also found [this article](https://ross-hagan.com/blog/instrument-nextjs-opentelemetry) which is following a different approach, but it stopped working for me even in local environment

**Comment by dannyDNS** [comment]
This is still an issue.

**Comment by Aaqu** [comment]
> This is still an issue.
> 
> It seems that if you disable `swcMinify` in `next.config.js`, instrumentation works with standalone builds.

`swcMinify: false` not work for me, instrumentation starts when open `http://localhost:3000`

**Comment by wisdomstar94** [comment]
I'm having the same problem.
When running with standalone, the register function is not called when it boots. :(

**Comment by h0wXD** [comment]
Ran into the same problem with same behavior as @Aaqu reported. Using Next.js v13.5.4

**Comment by evanshortiss** [comment]
This bug is also present in 14.0.1. Works fine with `next dev`, but `next build && node .next/standalone/server.js` fails to call the `register` function until a request is received.

**Comment by Starefossen** [comment]
@jankaifer @sophiebits @feedthejim @huozhi this is still broken event though #48615 was merged.

**Comment by rchubatov** [comment]
> Thanks @tomsseisums, I can confirm that it runs when handling a request. When I read the documentation, I thought there is only one server instance. So is this expected behavior?
> 
> > If you export a function named register from this file, we will call that function whenever a new Next.js server instance is bootstrapped. When your register function is deployed, it will be called on each cold boot (but exactly once in each environment).
> > https://nextjs.org/docs/app/building-your-application/optimizing/instrumentation

----

There seems to be a bug as the correct behavior should be (from the docs):
> When your register function is deployed, it will be called on each cold boot (**but exactly once in each environment**).

A simple workaround is to create some flag variable (or several) in **instrumentation.ts** and check if the `register` function has been called:

`
let isHookRegisteredInNode = false;
let isHookRegisteredInEdge = false;

export async function register() {
  if (process.env.NEXT_RUNTIME === 'nodejs' && isHookRegisteredInNode === false) {
    // to do something...
    isHookRegisteredInNode = true;
  }
  
  if (process.env.NEXT_RUNTIME === 'edge' && isHookRegisteredInEdge === false) {
    // to do something...
    isHookRegisteredInEdge = true;
  }
}
`

This way `register` will be called only once per environment on the first request. And when the bug is fixed, our code will still work.

**Comment by aNyMoRe0505** [comment]
Hello everyone, I'd like to ask if you're experiencing this issue only in standalone mode.
I'm using version 14.0.3, and even with the normal `next build && next start`, I encounter the same problem.
The register function is only called after receiving the request but works fine with `next dev`
Is anyone else facing similar issues?

**Comment by paschaldev** [comment]
@aNyMoRe0505 I'm experiencing the same issue. I chose to start a Queue worker using the `instrumentation.ts` file, but this doesn't work after deployment. The `register` is not called at all even when I make a request.

Everything works smoothly locally

**Comment by Aaqu** [comment]
> Any update on this? I'm experiencing the exact same issue. The only workaround is to not use the standalone export mode.

vercel will not sell if it works in standalone mode...

**Comment by daveyjones** [comment]
I've created a separate issue for those experiencing similar issues without `standalone` mode enabled.

https://github.com/vercel/next.js/issues/59999

**Comment by AntonAndreevichMoroz** [comment]
I am using the latest version 14.0.5-canary.58 and am also facing this problem. The instrumentation call after assembly is not called when the server starts, but when the route is first called. In dev mode, as expected, instrumentation exits when the server starts.

**Comment by Aaqu** [comment]
Any updates on this, please?

**Comment by Aaqu** [comment]
still the same problem in `Next.js 14.1.1-canary.52`

**Comment by Aaqu** [comment]
instrumentation seems to work in Next.js 14.2.0-canary.44 🎉 more in #63536

next.config.mjs
```
const nextConfig = {
  output: "standalone",
  compiler: {
    removeConsole: false,
  },
  experimental: {
    instrumentationHook: true,
  }
};
```

src/app/api/boot/boot.ts
```
const bootedServices = {
  service1: false,
};

export const bootHandler = async () => {
  if (!bootedServices.service1) {
    console.log("🐸 api/boot => SERVICE1");
    bootedServices.service1 = true;

    // your code here
  }
 return bootedServices;
};
```

src/app/api/boot/route.js
```
export async function GET() {
  const bootedServices = await bootHandler();
  return NextResponse.json({ bootedServices });
}

export const dynamic = "force-dynamic";
```

src/instrumentation.ts
```
export async function register() {
  if (process.env.NEXT_RUNTIME === 'nodejs') {
    console.log('nodejs');
    setTimeout(async () => {
      await fetch('http://0.0.0.0:3000/api/boot', {cache:"no-cache"});
    }, 3000);
  }
}
```

*tested in docker with file `.npmrc` includs `node-linker=hoisted`
**code inspiration [roneymaia](https://github.com/roneymaia) -> [code](https://github.com/vercel/next.js/discussions/15341#discussioncomment-5700982)

**Comment by aemr3** [comment]
It's **not working** for us **still** in `Next.js 14.2.3`.  Instrumentation hook is not included in `standalone` output.

**Comment by Aaqu** [comment]
> It's **not working** for us **still** in `Next.js 14.2.3`. Instrumentation hook is not included in `standalone` output. It's very strange that this issue is still open. It's been 1 year:)

You have a example? I test my code from [instrumentation](https://github.com/vercel/next.js/issues/49897#issuecomment-2022510320), work on Next.js 14.2.3 with docker

**Comment by leonidasqq** [comment]
> You have a example? I test my code from [instrumentation](https://github.com/vercel/next.js/issues/49897#issuecomment-2022510320), work on Next.js 14.2.3 with docker

It turned out that we had some issues with opentelemetry library. Your example worked for us. We can confirm that latest version is working with standalone output and instrumentation 🚀

**Comment by vpalos** [comment]
We had the same problem on version `14.2.3` and to make it work we had to **move `instrumentation.ts` inside the `src` folder** (if the app was configured to use it at the start).

**Comment by feedthejim** [comment]
I'm gonna close this as it seems to be mostly fixed.
